File: Review-14-Rust-Builder-Pattern/Example-1/src/room.rs
use crate::error::*;
use crate::flooring::*;
use std::cmp::Ordering;
use std::fmt;
use std::fmt::Display; //,Formatter,Result};

#[derive(Clone)]
pub struct DimensionSet {
    pub length: f64,
    pub width: f64,
}

impl DimensionSet {
    pub fn new(l: f64, w: f64) -> Self {
        DimensionSet {
            length: l,
            width: w,
        }
    }
}

impl Default for DimensionSet {
    fn default() -> Self {
        DimensionSet::new(1f64, 1f64)
    }
}

impl From<(f64, f64)> for DimensionSet {
    fn from(dims: (f64, f64)) -> Self {
        DimensionSet::new(dims.0, dims.1)
    }
}

#[derive(Clone)]
pub struct Room {
    pub name: String,
    pub dimensions: DimensionSet,
    pub flooring: Flooring,
}

impl Room {
    /// Compute the area of flooring for a room.
    pub fn area(&self) -> f64 {
        self.dimensions.width * self.dimensions.length
    }

    /// Compute the flooring cost based on `self.area()` and unit cost.
    pub fn flooring_cost(&self) -> f64 {
        self.area() * self.flooring.unit_cost
    }
}

impl Default for Room {
    fn default() -> Self {
        Room {
            name: "Generic".to_string(),
            dimensions: Default::default(),
            flooring: Default::default(),
        }
    }
}

impl Display for Room {
    #[allow(unused_must_use)]
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "Room ({})", self.name);
        writeln!(f, "  {:<6}: {:>8.1}", "Length", self.dimensions.length);
        writeln!(f, "  {:<6}: {:>8.1}", "Width", self.dimensions.width);
        writeln!(f, "  {:<6}: {:>8.1}", "Area", self.area());
        writeln!(f);
        writeln!(f, "  Flooring  : {}", self.flooring.type_name);
        writeln!(f, "  Unit Cost : $ {:>8.2}", self.flooring.unit_cost);
        writeln!(f, "  Total Cost: $ {:>8.2}", self.flooring_cost())
    }
}

impl PartialOrd for Room {
    fn partial_cmp(&self, rhs: &Self) -> Option<Ordering> {
        if self.name.eq(&rhs.name) {
            self.area().partial_cmp(&rhs.area())
        } else {
            self.name.partial_cmp(&rhs.name)
        }
    }
}

impl PartialEq for Room {
    fn eq(&self, rhs: &Self) -> bool {
        return self.name.eq(&rhs.name) && self.area().eq(&rhs.area());
    }
}

//------------------------------------------------------------------------------
pub struct RoomBuilder {
    name: Option<String>,
    length: Option<f64>,
    width: Option<f64>,
    flooring: Option<Flooring>,
}

impl RoomBuilder {
    pub fn new() -> Self {
        RoomBuilder {
            name: None,
            length: None,
            width: None,
            flooring: None,
        }
    }

    pub fn from_existing(self, room: &Room) -> Self {
        self.with_name(&room.name.clone())
            .with_flooring(room.flooring.clone())
            .with_dimensions(room.dimensions.length, room.dimensions.width)
    }

    pub fn with_name(mut self, nme: &str) -> Self {
        self.name = Some(nme.to_owned());

        self
    }

    pub fn with_flooring(mut self, flooring: Flooring) -> Self {
        self.flooring = Some(flooring);

        self
    }

    pub fn with_dimensions(mut self, l: f64, w: f64) -> Self {
        self.length = Some(l);
        self.width = Some(w);

        self
    }

    pub fn build(self) -> Result<Room, BuildError> {
        if self.name.is_none() {
            return Err(BuildError::GenericError("Name can not be blank"));
        }

        let room = Room {
            name: self.name.unwrap(),
            dimensions: DimensionSet::new(self.length.unwrap(), self.width.unwrap()),
            flooring: self.flooring.unwrap(),
        };

        Ok(room)
    }
}

File: Review-14-Python-Builder-Pattern/Example-2/room.py
from __future__ import annotations

import copy
from dataclasses import dataclass, field
from typing import Never, Optional, Self


@dataclass
class DimensionSet:
    length: float
    width: float


@dataclass
class Flooring:
    type_name: str
    unit_cost: float

    def update(self, nme: str, unit_c: float) -> None:
        """
        Replace/update the flooring.

        Args:
            nme: flooring type name
            unit_c: unit cost

        """

        self.type_name = nme
        self.unit_cost = unit_c


@dataclass
class Room:
    name: str
    dimensions: DimensionSet
    flooring: Flooring

    def area(self) -> float:
        """
        Compute the area of flooring for a room.
        """

        return self.dimensions.width * self.dimensions.length

    def flooring_cost(self) -> float:
        """
        Compute the flooring cost based on `self.area()` and unit cost.
        """

        return self.area() * self.flooring.unit_cost

    def __deepcopy__(self, memo) -> Room:
        cpy = Room(
            self.name, copy.deepcopy(self.dimensions), copy.deepcopy(self.flooring)
        )

        return cpy

    def __str__(self) -> str:
        return "\n".join(
            (
                f"Room ({self.name})",
                f"  Length: {self.dimensions.length:>8.1f}",
                f"  Width : {self.dimensions.width:>8.1f}",
                f"  Area  : {self.area():>8.1f}",
                "",
                f"  Flooring  : {self.flooring.type_name}",
                f"  Unit Cost : $ {self.flooring.unit_cost:>8.2f}",
                f"  Total Cost: $ {self.flooring_cost():>8.2f}",
                "\n",
            )
        )


class RoomBuilder:
    def __init__(self) -> None:
        self.__name: Optional[str] = None
        self.__length: Optional[float] = None
        self.__width: Optional[float] = None
        self.__flooring_type: Optional[str] = None
        self.__flooring_unit_cost: Optional[float] = None

    def from_template(self, room: Room) -> Self:
        self.__name = copy.deepcopy(room.name)
        self.__length = copy.deepcopy(room.dimensions.length)
        self.__width = copy.deepcopy(room.dimensions.width)
        self.__flooring_type = copy.deepcopy(room.flooring.type_name)
        self.__flooring_unit_cost = copy.deepcopy(room.flooring.unit_cost)

        return self

    def substitute_flooring(self, nme: str, unit_c: float) -> Self:
        self.with_flooring(nme, unit_c)

        return self

    def with_name(self, nme: str) -> Self:
        """
        Set the name using the builder pattern.

        Args:
            nme: room name
        """

        self.__name = nme.strip()

        return self

    def with_flooring(self, nme: str, unit_c: float) -> Self:
        """
        Set the Flooring using the builder pattern.

        Args
            nme: flooring type name
            unit_c: unit cost

        """

        self.__flooring_type = nme
        self.__flooring_unit_cost = unit_c

        return self

    def with_dimensions(self, l: float, w: float) -> Self:
        """
        Set the Flooring using the builder pattern.

        Args:
            l: length
            w: width

        """

        self.__length = l
        self.__width = w

        return self

    def __check_name(self, val: str | None, name: str) -> None:
        """
        Raise a Value Error if:
          1. val was not set
          2. val is the empty string
          3. val has fewer than 3 characters
        """

        if not val:
            raise ValueError(f'"{name}" was not set')

        if len(val) < 3:
            raise ValueError('"{name}" len("{val}") < 3')

    def __check_num(self, val: float | int, name: str) -> None:
        if val <= 0:
            raise ValueError(f'"{name}" <= 0')

    def build(self) -> Room:
        self.__check_name(self.__name, "name")
        self.__check_name(self.__flooring_type, "flooring type")

        # TODO: Duplicate check (remove)
        if not self.__flooring_type:
            raise ValueError("No flooring type was set")

        if not self.__flooring_unit_cost:
            raise ValueError("No flooring cost was set")

        if not self.__length:
            raise ValueError("No length was set")

        if not self.__width:
            raise ValueError("No width was set")

        self.__check_num(self.__flooring_unit_cost, "flooring cost")
        self.__check_num(self.__length, "length")
        self.__check_num(self.__width, "width")

        room = Room(
            self.__name,  # type: ignore
            DimensionSet(self.__length, self.__width),
            Flooring(self.__flooring_type, self.__flooring_unit_cost),
        )

        return room
